Add System Information HTTP Headers

ejoerns opened the pull request:
This implements passing system information upon the first HTTP request when accessing the URL of a streaming bundle.

This allows the server to collect more information about the device in order to create more verbose logs or to actually build deployment logic on this.

Besides some standard information that is configurable, custom information can be provided by the `system-info` handler.

This implements parts of concept #1114

jluebbe commented on docs/reference.rst:
```diff
@@ -298,6 +298,21 @@ For more information about using the streaming support of RAUC, refer to
+  * ``machine-id``: Enables sending the *machine-id* as ``MACHINE-ID`` header field.
+  * ``serial``: Enables sending the *system serial* as ``SERIAL`` header field.
+  * ``variant``: Enables sending the *variant* as ``VARIANT`` header field.
+  * ``transaction-id``: Enables sending the *variant* as ``TRANSACTION-ID`` header field.
```
```suggestion
  * ``transaction-id``: Enables sending the *transaction UUID* as ``TRANSACTION-ID`` header field.
```

jluebbe commented on src/nbd.c:
```diff
@@ -521,11 +522,16 @@ static void start_read(struct RaucNBDContext *ctx, struct RaucNBDTransfer *xfer)
 
-static struct curl_slist *gstrv_to_slist(const GStrv strv)
+/* Appends Gstrv elements to curl_slist.
+ * If curs_slist does not exist yet (NULL passed), it will be created. */
```
This should be explicit about who will own `initial_list`.

ejoerns replied:
Will extend doc

ejoerns commented on docs/advanced.rst:
```diff
@@ -478,6 +478,21 @@ round-trip-time (RTT) lead to longer installation times.
+simple logic and rollout handling on it.
+
+The actual information exposed to the server is configurable in the
+``system.conf`` section ``[server]``.
```
Yes, this was my second thought on this. I was unsure whether we would need more options later for phased rollouts, etc. But maybe we can even move/duplicate these options later if that should be required.

ejoerns commented on docs/reference.rst:
```diff
@@ -298,6 +298,21 @@ For more information about using the streaming support of RAUC, refer to
+  This option takes a list of information to send as HTTP header fields.
+  Supported values are:
+
+  * ``boot-id``: Enables sending the *boot_id* as ``BOOT-ID`` header field.
```
Sounds like a plan, yes.

jluebbe replied:
Now that I've looked into this a bit more, IETF has deprecated the `X-` prefix: https://www.rfc-editor.org/rfc/rfc6648

The recommendation seems to be to prefix them with a vendor name for disambiguation. So perhaps i.e. `RAUC-Boot-ID`?

ejoerns replied:
Will change to `RAUC-` instead of `X-RAUC-`.

ejoerns commented on include/config_file.h:
```diff
@@ -32,6 +32,15 @@ typedef enum {
+	gboolean variant;
+	gboolean uptime;
+} RConfigHTTPSendHeaders;
+
```
The idea was to be more performant with integer comparison rather than requiring string comparison. Maybe it's not that relevant here. I have changed it to use string lists..

ejoerns commented:
Updated to address parts of the review:

* Added `X-RAUC-` prefix to headers
* use string lists for storing headers types
* added minimal config tests
* updated documentation

jluebbe commented on src/install.c:
```diff
@@ -1159,6 +1159,85 @@ static gboolean launch_and_wait_default_handler(RaucInstallArgs *args, gchar* bu
+	gchar *contents = NULL;
+	g_autoptr(GError) ierror = NULL;
+
+	if (!g_file_get_contents("/proc/uptime", &contents, NULL, &ierror)) {
```
`/proc/uptime` contains two values, see `proc(5)`:
```
/proc/uptime
       This  file  contains  two numbers (values in seconds): the uptime of the
       system (including time spent in suspend) and the amount of time spent in
       the idle process.
```
Did you intentionally use both?

ejoerns replied:
Not intentionally. Will limit to using first one only

jluebbe commented on src/install.c:
```diff
@@ -1159,6 +1159,85 @@ static gboolean launch_and_wait_default_handler(RaucInstallArgs *args, gchar* bu
+
+static gchar **assemble_info_headers(RaucInstallArgs *args)
+{
+	GPtrArray *headers = g_ptr_array_new_full(1, g_free);
```
`_new_full` with a fixed, low size isn't really useful. Perhaps just use `g_ptr_array_new_with_free_func`, as this function isn't called often.

jluebbe commented on src/install.c:
```diff
@@ -1159,6 +1159,85 @@ static gboolean launch_and_wait_default_handler(RaucInstallArgs *args, gchar* bu
+static gchar **assemble_info_headers(RaucInstallArgs *args)
+{
+	GPtrArray *headers = g_ptr_array_new_full(1, g_free);
+	GHashTableIter iter;
+	gchar *key = NULL;
+	gchar *value = NULL;
```
Move these into the `if (r_context()->system_info)` block where they are used.

ejoerns replied:
Ack.

jluebbe commented on docs/reference.rst:
```diff
@@ -298,6 +298,17 @@ For more information about using the streaming support of RAUC, refer to
   used instead of the system wide store of trusted TLS/HTTPS certificates.
 
+``send-headers``
+  This option takes a list of information to send as HTTP header fields.
```
This should mention `;` as the list element separator.

ejoerns replied:
Yes, will add

a3f commented on docs/reference.rst:
```diff
@@ -298,6 +298,17 @@ For more information about using the streaming support of RAUC, refer to
+  Supported values are:
+
+  * ``boot-id``: Enables sending the *boot_id* as ``X-RAUC-BOOT-ID`` header field.
+  * ``machine-id``: Enables sending the *machine-id* as ``X-RAUC-MACHINE-ID`` header field.
```
jfyi, [Man page says: ](https://www.freedesktop.org/software/systemd/man/machine-id.html):
> This ID uniquely identifies the host. It should be considered "confidential", and must not be exposed in untrusted environments, in particular on the network.

I am not sure if that warrants using `sd_id128_get_machine_app_specific` or if we can just expect the user not to download over HTTP if they care about this.

jluebbe replied:
We should point to the documentation and mention that this header should not be used over unauthenticated connections or if the server is not trusted. In most cases, I expect that the update server needs to know the real `machine-id` tough. If needed, we could add `*-app-specific` variants to `boot-id` and `machine-id` later.

ejoerns replied:
Will add a note for this option.

ejoerns commented:
Fixed leaking `enabled_headers` revealed by sanitizer CI check.

ejoerns commented on src/install.c:
```diff
@@ -1159,6 +1159,99 @@
+
+	return (gchar**) g_ptr_array_free(headers, FALSE);
+}
+
```
FTR: this references `do_install_bundle()` which existed before.

ejoerns commented:
Rebased onto master. No further changes.

jluebbe commented on docs/reference.rst:
```diff
@@ -389,6 +407,16 @@ See details about using handlers in `Custom Handlers (Interface)`_.
+  forwarded to the server upon RAUC's first streaming request.
+  In order to define forwarded info, this must be returned as a key prefixed.
+  with ``RAUC_HTTP_``.
+  The generated header field will be name of the key (with out the prefix)
+  where an ``X-RAUC-`` is prepended and all underscores are converted to
+  hyphens.
+  E.g. ``RAUC_HTTP_MY_CUSTOM_INFO=dummyvalue`` will emit a header
+  ``X-RAUC-MY-CUSTOM-INFO: dummyvalue``.
```
```suggestion
  The generated header field will be name of the key (with out the prefix)
  where an ``RAUC-`` is prepended and all underscores are converted to
  hyphens.
  E.g. ``RAUC_HTTP_MY_CUSTOM_INFO=dummyvalue`` will emit a header
  ``RAUC-MY-CUSTOM-INFO: dummyvalue``.
```

jluebbe commented on src/install.c:
```diff
@@ -1158,6 +1158,99 @@ static gboolean launch_and_wait_default_handler(RaucInstallArgs *args, gchar* bu
+
+	/* hand char referenced by uptime[0] over to total_time to prevent autofreeing */
+	total_time = uptime[0];
+	uptime[0] = NULL;
```
Will uptime[1] be freed this way? `g_strfreev()` docs talk about a NULL-terminated array. Simply use `g_strdup()`.

ejoerns replied:
Ack

jluebbe commented:
> Note: Preview for high-level feedback. Still lacks some details.

Is this still accurate or can we merge this?

ejoerns commented:
> > Note: Preview for high-level feedback. Still lacks some details.
> 
> Is this still accurate or can we merge this?

Not valid any more. I've removed the comment. Can be merged as soon as we have the next point release ready.
